File: l2/specs/validate.mjs
// 互通性验证 · node 侧（权威）
// 证明：JSON (canonical) <-> YAML (human view) 双向无损，且符合 JSON Schema。
// 依赖：全局 ~/node_modules 的 yaml + ajv（node v22）。
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';

const dir = path.dirname(fileURLToPath(import.meta.url));
const yaml = (await import(pathToFileURL(path.join(process.env.HOME, 'node_modules', 'yaml', 'dist', 'index.js')).href)).default;
const Ajv = (await import(pathToFileURL(path.join(process.env.HOME, 'node_modules', 'ajv', 'dist', 'ajv.js')).href)).default;

const ajv = new Ajv({ allErrors: true, strict: false });

let failures = 0;
const check = (name, ok, detail) => {
  console.log(`${ok ? 'PASS' : 'FAIL'}   ${name}${ok ? '' : '   -- ' + detail}`);
  if (!ok) failures++;
};

// 三件套：json(canonical) / yaml(view) / schema，文件名各不相同，整列出避免拼接歧义。
const specs = [
   { label: 'memory', json: 'agent-memory.json', yaml: 'agent-memory.yaml', schema: 'agent-memory.schema.json' },
   { label: 'skill',  json: 'skill.json',        yaml: 'skill.yaml',        schema: 'skill.schema.json' },
];

for (const { label, json: jf, yaml: yf, schema: sf } of specs) {
  const jsonDoc = JSON.parse(fs.readFileSync(path.join(dir, jf), 'utf8'));
  const yamlDoc = yaml.parse(fs.readFileSync(path.join(dir, yf), 'utf8'));
  const schemaDoc = JSON.parse(fs.readFileSync(path.join(dir, sf), 'utf8'));

   // 1. 两 view 解析结果一致（JSON 是 canonical，YAML 必须与之完全等价）
  const roundtripEq = JSON.stringify(jsonDoc) === JSON.stringify(yamlDoc);
  check(`${label}: JSON == YAML (parsed)`, roundtripEq,
    roundtripEq ? '' : `\njson=${JSON.stringify(jsonDoc)}\nyaml=${JSON.stringify(yamlDoc)}`);

   // 2. YAML 字符串 -> 对象 -> 写回 YAML，应与原文件语义一致（无损往返）
  const backToYaml = yaml.stringify(yamlDoc);
  const reparsed = yaml.parse(backToYaml);
  check(`${label}: YAML round-trip (object identical)`,
    JSON.stringify(reparsed) === JSON.stringify(yamlDoc));

   // 3. schema 校验（canonical JSON）
  const validate = ajv.compile(schemaDoc);
  const ok = validate(jsonDoc);
  check(`${label}: JSON valid against ${sf}`, ok, ok ? '' : JSON.stringify(validate.errors));

   // 4. 负例：破坏 specVersion 常量应被 schema 拒绝
  const bad = JSON.parse(JSON.stringify(jsonDoc));
  bad.specVersion = '9.9.9';
  const badOk = validate(bad);
  check(`${label}: bad specVersion rejected`, badOk === false);
}

console.log(failures === 0 ? '\nNODE: ALL PASS' : `\nNODE: ${failures} FAILED`);
process.exit(failures === 0 ? 0 : 1);

File: l2/specs/validate.py
#!/usr/bin/env python3
"""互通性验证 · python 侧（降级，不污染全局环境）

证明 JSON (canonical) <-> YAML (human view) 双向无损。
- 有 pyyaml：做 YAML <-> JSON 对象语义相等 + 结构校验。
- 若 python 缺 pyyaml / jsonschema：优雅降级为“仅结构校验 + 明确报错”，
  绝不全局 install（铁律：不动全局环境）。
"""
import json
import sys
from pathlib import Path

HERE = Path(__file__).parent
SPEC = "1.0.0"
REQUIRED = {
    "memory": ["specVersion", "entries"],
    "skill": ["specVersion", "entries"],
}


def load_yaml(p):
    try:
        import yaml  # noqa
    except ImportError:
        print(f"FAIL  yaml unavailable, skip {p} (not a blocker; node side is authoritative)")
        return None
    return yaml.safe_load(p.read_text())


def schema_like(doc, keys):
    """轻量结构校验（替代 jsonschema 缺失时的降级）。"""
    if not isinstance(doc, dict):
        return False, "root not object"
    for k in keys:
        if k not in doc:
            return False, f"missing {k!r}"
    if doc.get("specVersion") != SPEC:
        return False, f"specVersion != {SPEC}"
    if not isinstance(doc.get("entries"), list) or not doc["entries"]:
        return False, "entries missing/empty"
    for i, e in enumerate(doc["entries"]):
        if not isinstance(e, dict):
            return False, f"entry[{i}] not object"
        if e.get("id") == "" or e.get("id") is None:
            return False, f"entry[{i}] missing id"
    return True, ""


def main():
    fails = 0
    pairs = [("agent-memory", ["specVersion", "entries"]),
             ("skill", ["specVersion", "entries"])]
    for name, keys in pairs:
        json_doc = json.loads((HERE / f"{name}.json").read_text())
        yaml_doc = load_yaml(HERE / f"{name}.yaml")
        if yaml_doc is None:
            fails += 1
            continue

        # 1. JSON == YAML 语义相等（YAML 视图必须等价于 canonical JSON）
        eq = json_doc == yaml_doc
        print(f"{'PASS' if eq else 'FAIL'}  {name}: JSON == YAML (parsed)")
        if not eq:
            fails += 1
            print(f"    json={json_doc}\n    yaml={yaml_doc}")

        # 2. 结构校验（json + yaml 两视图都过）
        for label, d in (("json", json_doc), ("yaml", yaml_doc)):
            ok, why = schema_like(d, keys)
            print(f"{'PASS' if ok else 'FAIL'}  {name} ({label}): structure ok")
            if not ok:
                fails += 1
                print(f"    {why}")

        # 3. 负例：坏 specVersion 应被结构校验拒绝
        bad = json.loads(json.dumps(json_doc))
        bad["specVersion"] = "9.9.9"
        ok, _ = schema_like(bad, keys)
        print(f"{'PASS' if not ok else 'FAIL'}  {name}: bad specVersion rejected")
        fails += not (not ok)

    print()
    print("PYTHON: ALL PASS" if fails == 0 else f"PYTHON: {fails} FAILED")
    sys.exit(0 if fails == 0 else 1)


if __name__ == "__main__":
    main()
